In [20]:
# Import 
import pandas as pd
import numpy as np
from scipy.stats import chi2_contingency, bootstrap
from statsmodels.stats.multitest import multipletests

In [21]:
# Data Loading
groups_df = pd.read_csv('data/groups.csv', sep=';')
group_add_df = pd.read_csv('data/group_add.csv')
checks_df = pd.read_csv('data/checks.csv', sep=';')

In [22]:
# Cohort Consolidation
groups_result_df = pd.concat([groups_df, group_add_df], ignore_index=True)

In [23]:
# Rename columns for consistency
checks_df = checks_df.rename(columns={'student_id': 'id'})

In [24]:
# A left join to ensure all randomized users remain in the dataset
experiment_df = pd.merge(groups_result_df, checks_df, on='id', how='left')

# Fill NaN revenue values with 0 for non-paying users
experiment_df['rev'] = experiment_df['rev'].fillna(0)

# Create a boolean conversion indicator
experiment_df['converted'] = (experiment_df['rev'] > 0).astype(int)

In [25]:
# Base metric calculation
# Calculate ARPU (all users), ARPPU (paying users only), and Conversion Rate
metrics = experiment_df.groupby('grp').agg(
    total_users=('id', 'count'),
    paying_users=('converted', 'sum'),
    arpu=('rev', 'mean'),
    median_rev_all=('rev', 'median')
)

metrics['conversion_rate'] = metrics['paying_users'] / metrics['total_users']

# Calculate ARPPU specifically for converted users
arppu_df = experiment_df[experiment_df['converted'] == 1].groupby('grp').agg(
    arppu=('rev', 'mean'),
    median_rev_paying=('rev', 'median')
)

final_metrics = pd.concat([metrics, arppu_df], axis=1)
final_metrics.round(4)

,total_users,paying_users,arpu,median_rev_all,conversion_rate,arppu,median_rev_paying
grp,,,,,,,
A,14693,107,6.2680,0.0,0.0073,860.7103,690.0
B,59883,434,8.0362,0.0,0.0072,1108.8249,935.0


### Justification for the Selection of Statistical Evaluation Methods

The evaluation utilizes statistical tests aligned with the mathematical properties of the target metrics. Conversion rate is a binary outcome mapping users to either a converted or non-converted state. Therefore, the Chi-Square test for independence is the correct method to evaluate differences in proportional frequencies across the mutually exclusive experimental groups.  Revenue metrics typically show continuous, non-normal distributions with a heavy right skew due to a large volume of zero-value observations and high-value outliers. Using standard parametric frameworks under these conditions increases the risk of false positive results. Bootstrap resampling via the percentile method resolves this by building an empirical sampling distribution directly from the observed data. This generates reliable confidence intervals that reflect the true variance in revenue without relying on invalid theoretical distribution assumptions.

In [26]:
# Statistical Testing - Conversion Rate ( Chi-Square)
# Test for conversion rate differences
contingency_table = [
    [metrics.loc['A', 'paying_users'], metrics.loc['A', 'total_users'] - metrics.loc['A', 'paying_users']],
    [metrics.loc['B', 'paying_users'], metrics.loc['B', 'total_users'] - metrics.loc['B', 'paying_users']]
]
chi2_stat, p_val_cr, dof, ex = chi2_contingency(contingency_table)
print(f"Conversion Rate Chi-Square p-value: {p_val_cr:.4f}")

Conversion Rate Chi-Square p-value: 1.0000


In [27]:
# Statistical Testing - ARPU (Bootstrap)
arpu_a = experiment_df.loc[experiment_df['grp'] == 'A', 'rev'].values
arpu_b = experiment_df.loc[experiment_df['grp'] == 'B', 'rev'].values

def stat_mean(a, b, axis=0):
    return np.mean(b, axis=axis) - np.mean(a, axis=axis)

res_arpu = bootstrap(
    data=(arpu_a, arpu_b),
    statistic=stat_mean,
    paired=False,
    vectorized=True,
    confidence_level=0.95,
    n_resamples=10000,
    method='percentile'
)
print(f"ARPU Difference 95% CI: [{res_arpu.confidence_interval.low:.4f}, {res_arpu.confidence_interval.high:.4f}]")

ARPU Difference 95% CI: [-0.1464, 3.5922]


In [28]:
# Statistical Testing - ARPPU (Bootstrap)
arppu_a = experiment_df.loc[(experiment_df['grp'] == 'A') & (experiment_df['converted'] == 1), 'rev'].values
arppu_b = experiment_df.loc[(experiment_df['grp'] == 'B') & (experiment_df['converted'] == 1), 'rev'].values

res_arppu = bootstrap(
    data=(arppu_a, arppu_b),
    statistic=stat_mean,
    paired=False,
    vectorized=True,
    confidence_level=0.95,
    n_resamples=10000,
    method='percentile'
)
print(f"ARPPU Difference 95% CI: [{res_arppu.confidence_interval.low:.4f}, {res_arppu.confidence_interval.high:.4f}]")

ARPPU Difference 95% CI: [74.4360, 411.4591]


### Conclusions Based on the Analysis of the Assessment Results

The analysis demonstrates a clear distinction between user acquisition and transaction value. The Chi-Square test for the conversion rate resulted in a p-value of 1.0000, indicating that the new payment flow had absolutely no impact on a user's likelihood to complete a purchase. The treatment did not increase the overall proportion of paying users.

However, the revenue metrics reveal a specific, localized financial impact. The 95% confidence interval for ARPPU is [74.4360, 411.4591], which strictly excludes zero. This confirms a statistically significant increase in the revenue generated by the specific subset of users who successfully completed a payment. The new flow effectively increases the transaction size for customers who already intend to buy.

Conversely, the 95% confidence interval for ARPU is [-0.1464, 3.5922], which crosses zero. This lack of statistical significance at the overall ARPU level highlights a volume limitation. The additional revenue extracted from the paying subset is mathematically insufficient to significantly increase the average revenue across the entire assigned user base.

### Business Recommendation:

The new payment mechanics act as an optimization for existing spenders but fail to expand the broader revenue base. Because the intervention does not increase overall average revenue per user, the new payment flow should not be rolled out if the primary strategic goal is top-line revenue growth across the entire customer base.